packages/worker/src/workflow/vm-shared.ts

```diff
@@ -0,0 +1,364 @@
+    // dump the error information to stderr and abort.
+    console.error('Unhandled rejection', { runId }, err);
+    process.exit(1);
+  });
+}
+/**
+ * Variant of {@link cutoffStackTrace} that works with FileLocation, keep this in sync with the original implementation
+ */
+function cutoffStructuredStackTrace(stackTrace: FileLocation[]): void {
+  stackTrace.shift();
+  if (stackTrace[0].functionName === 'initAll' && stackTrace[0].filePath === 'node:internal/promise_hooks') {
+    stackTrace.shift();
+  }
+  const idx = stackTrace.findIndex(({ functionName, filePath }) => {
+    return (
+      functionName &&
+      filePath &&
+      ((/^Activator\.\S+NextHandler$/.test(functionName) &&
+        /.*[\\/]workflow[\\/](?:src|lib)[\\/]internals\.[jt]s$/.test(filePath)) ||
+        (/Script\.runInContext/.test(functionName) && /^node:vm|vm\.js$/.test(filePath)))
+    );
+  });
+  if (idx > -1) {
+    stackTrace.splice(idx);
+  }
+}
+
+function getActivator(promise: Promise<any>): Activator | undefined {
+  // Access the global scope associated with the promise (unique per workflow - vm.context)
+  // See for reference https://github.com/patriksimek/vm2/issues/32
+  const ctor = promise.constructor.constructor;
+  return ctor('return globalThis.__TEMPORAL_ACTIVATOR__')();
+}
+
+/**
+ * Internal helper to format callsite "name" portion in stack trace
+ */
+function formatCallsiteName(callsite: NodeJS.CallSite): string | null {
+  const typeName = callsite.getTypeName();
+  const methodName = callsite.getMethodName();
+  const functionName = callsite.getFunctionName();
+  const isConstructor = callsite.isConstructor();
+
+  return typeName && methodName
+    ? `${typeName}.${methodName}`
+    : isConstructor && functionName
+    ? `new ${functionName}`
+    : functionName;
+}
+
+/**
+ * Global handlers for overriding stack trace preparation and promise hooks
+ */
+export class GlobalHandlers {
+  currentStackTrace: FileLocation[] | undefined = undefined;
+  bundleFilenameToSourceMapConsumer = new Map<string, SourceMapConsumer>();
+  origPrepareStackTrace = Error.prepareStackTrace;
+  private stopPromiseHook = () => {
+    // noop
+  };
+  installed = false;
+
+  async addWorkflowBundle(workflowBundle: WorkflowBundleWithSourceMapAndFilename): Promise<void> {
+    const sourceMapConsumer = await new SourceMapConsumer(workflowBundle.sourceMap);
+    this.bundleFilenameToSourceMapConsumer.set(workflowBundle.filename, sourceMapConsumer);
+  }
+
+  removeWorkflowBundle(workflowBundle: WorkflowBundleWithSourceMapAndFilename): void {
+    this.bundleFilenameToSourceMapConsumer.delete(workflowBundle.filename);
+  }
+
+  /**
+   * Set the global hooks, this method is idempotent
+   */
+  install(): void {
+    if (!this.installed) {
+      this.overridePrepareStackTrace();
+      this.setPromiseHook();
+      this.installed = true;
+    }
+  }
+
+  /**
+   * Unset all installed global hooks
+   * TODO(bergundy): Where should this be called?
```

Comment:
Refcount workers/runtimes?

The process-wide handlers I think are another good example of why we're really more like a process that hosts user code than a library they plug into their app. But, obviously that's not gonna change in this PR.

Reply:
It's not super critical since we typically install these in a separate thread which is cleaned up after worker shutdown.
The only case that isn't covered is debug mode where it's insignificant.
I'll add a comment in the code.